Better support for mod chests and items, shortcuts improvement

cpw opened the pull request:
A lot of this will merge into master as well I think but I don't have an MC1.0 env set up, and obviously I can't test anything with MC1.0 since nothing is out for that yet.

Hope it helps you a bit

cpw committed:
Added some stack trace output so we can see what went wrong with sorting

cpw committed:
Support for IronChests IC2 mod

cpw committed:
Fix up sorting for chests >9 columns wide

cpw committed:
Fixup for shift clicking stuff between custom containers and inventory
Also support for "one stack" type operations with a keybind

cpw committed:
Autogenerate items rather than putting them all into a single giant bucket

cpw committed:
Merge branch 'autogenerateitems' into mc1.8

cpw commented:
From previous issue:
1. Support IronChests mod (Gold and Diamond chests)
2. Fix up the sorting so that it works for chests >9 wide. (This touched a bunch of places to fix up stuff. Had to add in a parameter to a method to get it right).
3. Dummy items. I don't like that "it's not default so it all goes in one bucket" behaviour. I added some stuff so that if the mod didn't know about an item, it added it to the root of the tree, which I think is a nice thing to do.
4. Fixed up 1 stack handling. Your code didn't seem to handle it at all, which was disappointing. I think you were anticipating passing through to MC, but you don't let that happen either.

mkalam-alami commented on src/InvTweaksItemTree.java:
```diff
@@ -171,11 +171,19 @@ public class InvTweaksItemTree {
-            log.warning("Unknown item id: " + id);
-            return defaultItems;
+            // Generate a dummy item at the root for sorting 
+            InvTweaksItemTreeItem newItem=new InvTweaksItemTreeItem(String.format("UnknownItem-%d-%d",id,damage),id,damage,id*16+damage);
```
I approve of this, good idea.

cpw replied:
Just to note you need both parts of this file change: most mods use 1 item id and a bunch of damage values, so each damage value is normally treated as a different item. (Which the other part of the diff handles- it seems to work so far).
I find it really handy for mod items though- with redpower ic2 bc2 harvestcraft and others, I'll be editing your nice little file for months. Being able to just add "key" item groupings to it and let everything else fall into root category makes life a LOT easier :)

mkalam-alami commented:
Well since you're not dealing with obfuscated code, there's no reason for it not to work. Thanks, I'll merge and test this this week-end.

I'll make sure to mention you when releasing the next update ; would you like for me to add a link to your personal website/blog/mod if you have any? It might both give you some visibility and encourage people to contribute for the mod!

Thanks again for your work =)

cpw commented:
Heh, thanks. I don't really have a website or anything- I just love fixing bugs, making things "work" :)

You'll see my name on a few mods, I'm cpw at the mc forums too.

The merge went fine except that the ModCompatibility java was a bit screwy and mvn build failed on the "merged" code. As I said, I haven't had time to hack on it with MC1.0 (and MC1 is a bit pointless for a lot of these "fixes" since they're all mod compatibility stuff, which is still weeks out probably).

Anyway, thanks for accepting the tweaks. Enjoy!

cpw committed:
Fix a small issue: MAX_CONTAINER_SIZE is too small for really big chests.

Probably can be smaller than 200 though, but need to check, (though multipage chest springs to mind here too)

cpw committed:
Make armor sorting conditional

If you middle click on the player inventory area with a chest open you get NPE otherwise because the Armor slots aren't there with a chest open.
